keps/sig-storage/596-csi-inline-volumes/README.md

```diff
@@ -213,32 +250,214 @@ Inline volume requests can only participate in mount/unmount volume operations.
 * Single pod referencing two distinct inline volume request from the same driver: "should support multiple inline ephemeral volumes" in `ephemeral.go` (upcoming PR)
 * CSI Kubelet code invokes driver operations during mount for ephemeral volumes: `checkPodLogs` in `csi_mock_volume.go` (upcoming PR)
 * CSI Kubelet code invokes driver operation during unmount of ephemeral volumes: `checkPodLogs` in `csi_mock_volume.go` (upcoming PR)
 * CSI Kubelet cleans up ephemeral volume paths once pod goes away: TODO
-* Apply PodSecurity settings for allowed CSI drivers: TODO
 * Enable testing of an external ephemeral CSI driver: https://github.com/kubernetes/kubernetes/pull/79983/files#diff-e5fc8d9911130b421b74b1ebc273f458
 * Enable testing of the csi-host-path-driver in ephemeral mode in Kubernetes-CSI Prow jobs and Kubernetes itself: TODO
 
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind `CSIInlineVolume` feature flag
+- Initial unit tests and e2e tests implemented
+
+#### Beta
+
+- Feature flag enabled by default
+- CSI drivers can support both ephemeral inline volumes and persistent volumes
+
+#### GA
+
+- Remove dependency on deprecated `PodSecurityPolicy` and document new strategy
+- Upgrade / downgrade manual testing
+- Conformance tests implemented
+
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: CSIInlineVolume
+  - Components depending on the feature gate: kubelet
+
+###### Does enabling the feature change any default behavior?
+
+  Enabling the feature gate will **not** change the default behavior.
+  Each CSI driver must opt-in to support ephemeral inline volumes.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+  Yes. Disabling the feature gate will disable support for inline CSI volumes.
+  It can also be disabled on a specific CSI driver by removing the "Ephemeral"
+  VolumeLifecycleMode from `CSIDriver.VolumeLifecycleModes`.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+  Reenabling the feature will again allow CSI drivers that opt-in to be used
+  as ephemeral inline volumes in the pod spec.
+
+###### Are there any tests for feature enablement/disablement?
+
+  Yes, the unit tests will test with the feature gate enabled and disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+  A rollout should not impact running workloads, since the CSI driver
+  also needs to support ephemeral inline volumes, and the pod spec would need
+  to be updated to make use of the feature.
+
+  A rollback (or disabling the feature flag) may impact running workloads,
```

Comment:
This paragraph seems to be focused on "new workloads we're trying start".
What about the pods that use CSI inline volumes that are already running. Is there anything that could lead to their failures? Under normal circumstances? If they fails for some reason and needs to be restarted by Kubelet (e.g. due to Always restart policy)?

Reply:
I reworded the rollout section to (hopefully) clarify. I can't think of any reason why rollout would impact running pods, even in the case of pod restarts.

Reply:
OK - the previous paragraph now looks ok to me.

Regarding this one ("A rollback...") I would make it more explicit, that once these workloads are already running, they are fine. And it may be problematic if we want to start new ones or potentially we need to restart the running ones (if kubelet no longer has a FG enabled, then it will not call appropriate mount/unmount calls, right?

So (assuming I'm correct - please challenge me if I'm missing something), I would change to something like:
```
A rollback (or disabling the feature flag) will not impact already running workloads, however starting new ones (as well as potentially restarting ones that failed for some reason) would fail. Additionally, kubelet will fail to fully cleanup after pods that were running and taking advantage of inline volumes.
*It is highly recommended to delete any pods using this feature before disabling it*.
```

Does that make sense? Is that correct?

Reply:
It makes sense, with one caveat. If the feature flag is disabled, kubelet will still attempt to unmount existing volumes when requested to do so (i.e. there is no feature flag check when unmounting):
https://github.com/kubernetes/kubernetes/blob/66445662ad7b65ffdb144b9665608f7b6a4b57a6/pkg/volume/csi/csi_mounter.go#L380-L382
But it will not attempt to mount new volumes using this feature if it's disabled.
https://github.com/kubernetes/kubernetes/blob/66445662ad7b65ffdb144b9665608f7b6a4b57a6/pkg/volume/csi/csi_mounter.go#L380-L382
I changed your wording slightly to "kubelet may fail to fully cleanup" since kubelet will request `NodeUnpublishVolume` but cleanup then depends on the individual CSI driver, and we still recommend deleting pods using this feature before disabling it.